### DISCUSSION TITLE
Can anyone help me how to use this in colab along with local llms

### DISCUSSION THREAD

**Discussion by YaswanthDasamandam** [microsoft/autogen#89] [Q&A]
Hi , 
I normally use colab for using smaller Local models . But Here I am not able to create llm lite server and the autogen at the same time . 
because of inline . Can anyone help 

I read the Issue #46 here it was a discussion about llmlite server .. 

But can anyone create a colab for working with local llms . Then the project would be useful for everyone.

```python

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

torch.set_default_device("cuda")

model = AutoModelForCausalLM.from_pretrained("microsoft/phi-1_5", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained("microsoft/phi-1_5", trust_remote_code=True)
inputs = tokenizer('''```python
def print_prime(n):
   """
   Print all primes between 1 and n
   """''', return_tensors="pt", return_attention_mask=False)

outputs = model.generate(**inputs, max_length=200)
text = tokenizer.batch_decode(outputs)[0]

```

This would be very helpful for everyone

**Comment by YaswanthDasamandam** [comment]
If I am using the previous method to create what should I send as arguments here instead of the below code 

```python

        response = oai.ChatCompletion.create(
            context=messages[-1].pop("context", None), messages=self._oai_system_message + messages, **llm_config
        )
        return True, oai.ChatCompletion.extract_text_or_function_call(response)[0]
```

**Comment by LeoLjl** [comment]
Hi @YaswanthDasamandam, could you please check this out? https://microsoft.github.io/autogen/blog/2023/07/14/Local-LLMs

This method requires colab pro, as running commands in terminal is a colab pro feature.

**Reply by YaswanthDasamandam** [reply]
Is it possible to modify code so user defined methods can also be used instead of llm  lite server ??

**Reply by YaswanthDasamandam** [reply]
First of all thank you for the response

**Reply by sonichi** [reply]
You can either override the [`generate_oai_reply`](https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent#generate_oai_reply) method or the [`create`](https://microsoft.github.io/autogen/docs/reference/oai/completion#create) function, depending on which is more convenient for you.

**Reply by YaswanthDasamandam** [reply]
Thank you for the response.
1. In future are there any plans of creating a general class for these kind of implementations ? 
2. Currently everything is revolved around OpenAI API , surely open models are impoving , especially from microsoft . The solution given in the blog post is just a temperory fix . If this project to be used for big companies or projects  or even small developers then it is not an efficient way . I have been giving a thought  about implementing a class . If I find any simple way to change I will definetely post it here .
3.Main issue with creating class is all the code that is written for API  need to be removed which makes it harder to make any changes without breaking the code . 
4.It would be great if there were plans for creating scheme or general class and be able to modify it :) 
Thank you making autogen public .

**Reply by sonichi** [reply]
Thanks for sharing your feedback. Please take a look at #95

**Comment by YaswanthDasamandam** [comment]
how can we use huggingface models with hugging face token

**Comment by kinthaiofficial** [comment]
AutoGen with local LLMs works best when you set up an OpenAI-compatible endpoint first, then point AutoGen at it. Here's a working setup for Colab + Ollama:

```python
# Install Ollama in Colab (requires GPU runtime)
import subprocess
subprocess.Popen(['ollama', 'serve'])  # after installing: !curl -fsSL https://ollama.com/install.sh | sh

# Pull your model
subprocess.run(['ollama', 'pull', 'llama3:8b'])

# Configure AutoGen to use local endpoint
config_list = [{
    "model": "llama3:8b",
    "base_url": "http://localhost:11434/v1",
    "api_key": "ollama",  # can be any string
    "api_type": "openai",  # important: tells autogen to use OpenAI-compatible API
}]

llm_config = {"config_list": config_list, "seed": 42}

assistant = autogen.AssistantAgent(
    name="assistant",
    llm_config=llm_config
)
```

A few things that trip people up:

**Tool calling support** — not all local models support function/tool calls reliably. Llama 3 8B handles it acceptably; Mistral 7B is less reliable. For multi-agent orchestration where agents need tool calls, use at least a 13B model or Llama 3.

**Context length** — Colab free tier has limited RAM. Use `num_ctx: 4096` in the Ollama Modelfile to limit context and avoid OOM.

**Temperature for coding agents** — local models benefit from lower temperature (0.1-0.3) for code tasks. Higher temperature makes them more creative but less accurate on structured tasks.

For the multi-agent coordination patterns that work well with local models: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons — the economic routing model that decides when to use local vs cloud: https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents

Are you using Colab for experimentation or trying to build a production-viable local agent setup?
